packages/eslint-plugin/src/rules/strict-boolean-expressions.ts

```diff
@@ -275,6 +285,103 @@ export default createRule<Options, MessageId>({
       if (assertedArgument != null) {
         traverseNode(assertedArgument, true);
       }
+      if (isArrayMethodCallWithPredicate(context, services, node)) {
+        const predicate = node.arguments.at(0);
+
+        if (predicate) {
+          checkArrayMethodCallPredicate(predicate);
+        }
+      }
+    }
+
+    /**
+     * Dedicated function to check array method predicate calls. Reports predicate
+     * arguments that don't return a boolean value.
+     *
+     * Ignores the `allow*` options and requires a boolean value.
+     */
+    function checkArrayMethodCallPredicate(
+      predicateNode: TSESTree.CallExpressionArgument,
+    ): void {
+      const isFunctionExpression = ASTUtils.isFunction(predicateNode);
+
+      // custom message for accidental `async` function expressions
+      if (isFunctionExpression && predicateNode.async) {
+        return context.report({
+          node: predicateNode,
+          messageId: 'predicateCannotBeAsync',
+        });
+      }
+
+      const type = checker.getApparentType(
+        services.getTypeAtLocation(predicateNode),
+      );
+
+      const signatures = type.getCallSignatures();
+
+      if (
+        signatures.every(signature =>
+          tsutils
+            .unionTypeParts(signature.getReturnType())
+            .every(isBooleanType),
+        )
+      ) {
+        return;
+      }
+
+      const canFix = isFunctionExpression && !predicateNode.returnType;
+
+      return context.report({
+        node: predicateNode,
+        messageId: 'predicateReturnsNonBoolean',
+        suggest: canFix
+          ? [
+              {
+                messageId: 'explicitBooleanReturnType',
+                fix: fixer => {
+                  const lastParam = predicateNode.params.at(-1);
+
+                  // zero parameters
+                  if (!lastParam) {
+                    const closingBracket = nullThrows(
+                      context.sourceCode.getFirstToken(
+                        predicateNode,
+                        token => token.value === ')',
+                      ),
+                      'function expression with no arguments must have a closing parenthesis.',
+                    );
+
+                    return fixer.insertTextAfterRange(
+                      closingBracket.range,
+                      ': boolean',
+                    );
+                  }
+
+                  const closingBracket = nullThrows(
+                    context.sourceCode.getTokenAfter(lastParam),
+                    'missing token following function parameters.',
+                  );
+
+                  // one or more parameters wrapped with parens
+                  if (closingBracket.value === ')') {
```

Comment:
There is this utility also if you wanted, but it's fine as-is also

https://github.com/typescript-eslint/typescript-eslint/blob/3f02687d964c1e7888e5db62256d56cd72c4f281/packages/eslint-plugin/src/util/misc.ts#L227-L234

Reply:
Nice! I've used this (though I had to make slight changes to it) along with refactoring my old code for the fixer a bit too.